playbooks/group_vars/promtail_servers/00-defaults.yaml

```
@@ -1,15 +1,33 @@
 # vim:ts=2:sw=2:et:ai:sts=2
 ---
 
-apt__packages_to_install__promtail:
-  - promtail
-apt__services_to_enable__promtail:
-  - promtail
-apt__services_to_start__promtail:
-  - promtail
-
-ferm__services__promtail:
-  promtail:
-    dport: 9080
-    proto: tcp
-    saddr: '{{ internal_networks }}'
+promtail__listen_port: 9080
+promtail__storage_dir: /var/lib/promtail
+
+# If true, install ferm configuration file.
+promtail__install_ferm_svc: '{{ ferm__enable | bool }}'
+# Optionally limit access to a list of IP addresses.
+promtail__ferm_allow_from: '{{ internal_networks }}'
+
+# List of Loki instances to push logs to.
+promtail__client_urls: []
+# Parameters for Loki clients.
+promtail__client_defaults: {}
+# Full push clients configuration block,
+# if set overrides `promtail__client_urls` and `promtail__client_defaults`.
+promtail__clients: |
+  {% set res = [] %}
+  {% for url in promtail__client_urls %}
+  {% set _ = res.append(promtail__client_defaults | combine(dict(url=url))) %}
+  {% endfor %}
+  {{- res -}}
+
+promtail__scrape_configs: |-
+  {{ query('vars', *(query('varnames', '^promtail__scrape_config__.*')
```

```suggestion
  {{ query('vars', *(query('varnames', '^promtail__scrape_configs__.*')
```

(I think this was an unintentional typo in Tina's original suggestion, since I couldn't find any example of varname aggregation query that didn't use the exact same variable name.)

---

no, in fact this one is intentional; mirroring what I did for `libvirt_guests`:

```yaml
vm_volumes: |-
  {{ query('vars', *(query('varnames', '^vm_volume__.*') | select | sort)) }}
vm_interfaces: |-
  {{ query('vars', *(query('varnames', '^vm_interface__.*') | select | sort)) }}
```

The reason for this is that these are not like all the other var aggregation templates we have: these take an item (a dict) are produce a list of items, so they are more "ergonomic" to use. Compare:

```yaml
vm_interface__00_pubbr:
  type: bridge
  mac: '{{ pubnet_hwaddr }}'
  source:
    dev: vmbr0
  target:
    dev: vm-{{ vm_name }}-en0
```

With the alternative:

```yaml
vm_interfaces__00_pubbr:
- type: bridge
  mac: '{{ pubnet_hwaddr }}'
  source:
    dev: vmbr0
  target:
    dev: vm-{{ vm_name }}-en0
```

 In retrospect, I think I should have done the same for most aggregation of lists of complex items (but not for simple lists like `apt__packages_to_install`)